### PR TITLE
Add gallery of examples to documentation

### PR DESCRIPTION
**Pull request by mschmidt87**
This addresses one point in #174 : It adds a gallery showcasing our examples to the documentation. 

This is achieved in two steps:
- Add the necessary infrastructure (additional `sphinx-gallery` package, some changes in the documentation)
- Restructure example scripts to add comments that comply with the restructured Text formatting of `sphinx-gallery` (see here for a guide: https://sphinx-gallery.github.io/stable/syntax.html). Note that I dissolved the main function in some of the examples because it didn't seem necessary and without putting the main loop in a function, we can nicely break up the code in smaller chunks with docstrings in between.

To test this, execute the following:
```shell
cd docs
make html
firefox _build/html/index.html &
```
and navigate to `examples`. 

*Out of scope of this PR:*
- inclusion of figures from the examples: see comment in the issue #174 .

### CODE REVIEW

File: docs/index.rst [hunk 1]
```diff
@@ -1,4 +1,4 @@
-.. hal-cgp documentation master file, created by
+ .. hal-cgp documentation master file, created by
```
Comment (jakobj) [inline docs/index.rst:1]:
```suggestion
.. hal-cgp documentation master file, created by
```

File: examples/example_caching.py [hunk 2]
```diff
@@ -1,21 +1,34 @@
...
 second time and when you comment out the decorator on
 `inner_objective`."""
 
+import numpy as np
+import time
+
+import cgp
+
+# %%
+# We define the target function for regression in this examples.
```
Comment (jakobj) [inline examples/example_caching.py:17]:
```suggestion
# We define the target function for this example.
```

File: examples/example_caching.py [hunk 3]
```diff
@@ -1,21 +1,34 @@
...
 def f_target(x):
     return x ** 2 + x + 1.0
 
 
+# %%
+# We then define the objective function for the evolutionary
+# algorithm: It consists of an inner objective which we wrap with the
+# caching decorator that specified a pickle file that will be used for
+# the caching. This inner objective is then used by actual objective
+# function to update the fitness of the individual.
```
Comment (jakobj) [inline examples/example_caching.py:29]:
```suggestion
# We then define the objective function for the evolutionary
# algorithm: It consists of an inner objective which we wrap with the
# caching decorator. This decorator specifies a pickle file that will be used for
# caching results of fitness evaluations. The inner objective is then used by the objective
# function to compute (or retrieve from cache) the fitness of the individual.
```

File: examples/example_caching.py [hunk 4]
```diff
@@ -47,34 +60,38 @@ def objective(individual):
...
-
-    return pop.champion
-
-
-if __name__ == "__main__":
-    champion = evolution()
-    print(f"evolved function: {champion.to_sympy()}")
+# %%
+# Next, we define the parameters for the population, the genome of
+# individuals, the evolutionary algorithm, and the local search.
```
Comment (jakobj) [inline examples/example_caching.py:65]:
```suggestion
# individuals, and the evolutionary algorithm.
```

File: examples/example_caching.py [hunk 5]
```diff
@@ -47,34 +60,38 @@ def objective(individual):
...
+        "n_rows": 2,
+        "levels_back": 2,
+        "primitives": (cgp.Add, cgp.Sub, cgp.Mul, cgp.ConstantFloat),
+    },
+    "evolve_params": {"max_generations": 100, "min_fitness": -1e-12},
+}
+
+# %%
+# We then create the Population object and instantiate the local search
+# and evolutionary algorithms.
```
Comment (jakobj) [inline examples/example_caching.py:84]:
```suggestion
# We then create a Population instance and instantiate the evolutionary algorithm.
```

File: examples/example_caching.py [hunk 6]
```diff
@@ -47,34 +60,38 @@ def objective(individual):
...
+
+
+pop = cgp.Population(**params["population_params"], genome_params=params["genome_params"])
+ea = cgp.ea.MuPlusLambda(**params["ea_params"])
+
+# %%
+# Finally, we call the `evolve` method that performs the evolution.
```
Comment (jakobj) [inline examples/example_caching.py:91]:
```suggestion
# Finally, we call the `evolve` method to perform the evolutionary search.
```

File: examples/example_differential_evo_regression.py [hunk 7]
```diff
@@ -1,35 +1,50 @@
...
-import torch
-
-import cgp
-
+"""
+Example for differential evolutionary regression
+================================================
 
-"""Example demonstrating the use of Cartesian Genetic Programming for
+Example demonstrating the use of Cartesian Genetic Programming for
```
Comment (jakobj) [inline examples/example_differential_evo_regression.py:5]:
```suggestion
Example demonstrating the use of Cartesian genetic programming for
```

File: examples/example_differential_evo_regression.py [hunk 8]
```diff
@@ -1,35 +1,50 @@
...
+import functools
+import matplotlib.pyplot as plt
+import numpy as np
+import scipy.constants
+import torch
+
+import cgp
+
+# %%
+# We first define the target function that we want to regress.
```
Comment (jakobj) [inline examples/example_differential_evo_regression.py:32]:
```suggestion
# We first define the target function. Note that this function contains
# numeric values which are initially not available as constants to the search.
```

File: examples/example_differential_evo_regression.py [hunk 9]
```diff
@@ -1,35 +1,50 @@
...
+
 
 def f_target(x):
     return x[:, 0] ** 2 + 1.0 + np.pi
 
 
+# %%
+# Then we define the objective function for the evolution.  It
```
Comment (jakobj) [inline examples/example_differential_evo_regression.py:40]:
```suggestion
# Then we define the differentiable(!) objective function for the evolution.  It
```

File: examples/example_differential_evo_regression.py [hunk 10]
```diff
@@ -57,91 +72,104 @@ def objective(individual, seed):
...
-    genome_params = {
-        "n_inputs": 1,
-        "n_outputs": 1,
-        "n_columns": 20,
-        "n_rows": 1,
-        "levels_back": None,
-        "primitives": (cgp.Add, cgp.Sub, cgp.Mul, cgp.Parameter),
-    }
+# %%
+# We then create the Population object and instantiate the local search
```
Comment (jakobj) [inline examples/example_differential_evo_regression.py:106]:
```suggestion
# We then create a Population instance and instantiate the local search
```

File: examples/example_differential_evo_regression.py [hunk 11]
```diff
@@ -57,91 +72,104 @@ def objective(individual, seed):
...
-    obj = functools.partial(objective, seed=population_params["seed"])
 
-    cgp.evolve(
-        pop, obj, ea, **evolve_params, print_progress=True, callback=recording_callback,
-    )
+obj = functools.partial(objective, seed=population_params["seed"])
 
-    return history, pop.champion
+# %%
+# Finally, we call the `evolve` method that performs the evolution.
```
Comment (jakobj) [inline examples/example_differential_evo_regression.py:141]:
```suggestion
# Finally, we call the `evolve` method to perform the evolutionary search.
```

File: examples/example_evo_regression.py [hunk 12]
```diff
@@ -1,3 +1,12 @@
+"""
+Example for evolutionary regression
+===================================
+
+Example demonstrating the use of Cartesian Genetic Programming for
```
Comment (jakobj) [inline examples/example_evo_regression.py:5]:
```suggestion
Example demonstrating the use of Cartesian genetic programming for
```

File: examples/example_evo_regression.py [hunk 13]
```diff
@@ -6,9 +15,10 @@
 
 import cgp
 
-
-"""Example demonstrating the use of Cartesian Genetic Programming for
-two regression tasks."""
+# %%
+# We first define the target function that we want to regress. For
+# illustration purposes, we define two functions with different
+# difficulty to fit.
```
Comment (jakobj) [inline examples/example_evo_regression.py:21]:
```suggestion
# We first define target functions. For illustration purposes, we define two functions
# which present different levels of difficulty for the search. 
```

File: examples/example_evo_regression.py [hunk 14]
```diff
@@ -64,6 +81,19 @@ def objective(individual, target_function, seed):
...
 
 
+# %%
+# Next, we define the main loop of the evolution. To easily execute it
+# for different target functions, we wrap it into a function here. It
+# comprises:
+#
+# - defining the parameters for the population, the genome of individuals,
+#   and the evolutionary algorithm.
+# - creating the Population object and instantiate the local search and evolutionary algorithm.
```
Comment (jakobj) [inline examples/example_evo_regression.py:91]:
```suggestion
# - creating a Population instance and instantiate the evolutionary algorithm.
```

File: examples/example_evo_regression.py [hunk 15]
```diff
@@ -64,6 +81,19 @@ def objective(individual, target_function, seed):
...
+# - for bookkeeping of the results, we define the `history` dictionary and a recording_callback.
+#
+# Finally, we call the `evolve` method that performs the evolution.
```
Comment (jakobj) [inline examples/example_evo_regression.py:94]:
```suggestion
# Finally, we call the `evolve` method to perform the evolutionary search.
```

File: examples/example_mountain_car.py [hunk 16]
```diff
@@ -40,6 +46,13 @@ def __init__(self, idx, inputs):
         self._output = 10.0
 
 
+# %%
+# Then we define the objective function for the evolution.  The inner
+# objective which accepts a Python callable as input. The fitness of
+# the given callable on the task is then computed as the cumulative
+# reward over a fixed number of episodes.
```
Comment (jakobj) [inline examples/example_mountain_car.py:53]:
```suggestion
# Then we define the objective function for the evolution.  The inner
# objective accepts a Python callable as input. This callable determines the action taken by the agent upon receiving observations from the environment. The fitness of
# the given callable on the task is then computed as the cumulative
# reward over a fixed number of episodes.
```

File: examples/example_mountain_car.py [hunk 17]
```diff
@@ -97,6 +117,16 @@ def objective(ind, seed, n_total_steps):
     return ind
 
 
+# %%
+# We then define the main loop for the evolution, which consists of:
+#
+# - parameters for the population, the genome of individuals, and the evolutionary algorithm.
+# - creating the Population object and instantiate the local search and evolutionary algorithm.
```
Comment (jakobj) [inline examples/example_mountain_car.py:124]:
```suggestion
# - creating a Population instance and instantiate the evolutionary algorithm.
```

File: examples/example_mountain_car.py [hunk 18]
```diff
@@ -97,6 +117,16 @@ def objective(ind, seed, n_total_steps):
...
+# - for bookkeeping of the results, we define the `history` dictionary and a recording_callback.
+#
+# Finally, we call the `evolve` method that performs the evolution.
```
Comment (jakobj) [inline examples/example_mountain_car.py:127]:
```suggestion
# Finally, we call the `evolve` method to perform the evolutionary search.
```

File: examples/example_mountain_car.py [hunk 19]
```diff
@@ -198,6 +232,11 @@ def f(x):
     return cum_reward_all_episodes
 
 
+# %%
+# Furthermore, we define a function for an animated visualization for
+# the expressions that mark a jump in the fitness value.
```
Comment (jakobj) [inline examples/example_mountain_car.py:237]:
```suggestion
# Furthermore, we define a function for visualizing the agents behaviour for
# each expression that increase over the currently best performing individual.
```

File: examples/example_differential_evo_regression.py [hunk 20]
```diff
@@ -57,91 +72,104 @@ def objective(individual, seed):
...
 
-    # use an uneven number of gradient steps so they can not easily
-    # average out for clipped values
-    local_search_params = {"lr": 1e-3, "gradient_steps": 9}
+ea = cgp.ea.MuPlusLambda(**ea_params, local_search=local_search)
 
-    pop = cgp.Population(**population_params, genome_params=genome_params)
+# %%
+#  We define the `history` dictionary and a recording_callback for
+#  bookkeeping of the results.
```
Comment (jakobj) [inline examples/example_differential_evo_regression.py:125]:
```suggestion
#  We define a recording callback closure for bookkeeping of the progression of the evolution.
```

File: examples/example_evo_regression.py [hunk 21]
```diff
@@ -64,6 +81,19 @@ def objective(individual, target_function, seed):
...
 
+# %%
+# Next, we define the main loop of the evolution. To easily execute it
+# for different target functions, we wrap it into a function here. It
+# comprises:
+#
+# - defining the parameters for the population, the genome of individuals,
+#   and the evolutionary algorithm.
+# - creating the Population object and instantiate the local search and evolutionary algorithm.
+# - for bookkeeping of the results, we define the `history` dictionary and a recording_callback.
```
Comment (jakobj) [inline examples/example_evo_regression.py:92]:
```suggestion
# - we define a recording callback closure for bookkeeping of the progression of the evolution.
```

File: examples/example_mountain_car.py [hunk 22]
```diff
@@ -97,6 +117,16 @@ def objective(ind, seed, n_total_steps):
     return ind
 
 
+# %%
+# We then define the main loop for the evolution, which consists of:
+#
+# - parameters for the population, the genome of individuals, and the evolutionary algorithm.
+# - creating the Population object and instantiate the local search and evolutionary algorithm.
+# - for bookkeeping of the results, we define the `history` dictionary and a recording_callback.
```
Comment (jakobj) [inline examples/example_mountain_car.py:125]:
```suggestion
# - we define a recording callback closure for bookkeeping of the progression of the evolution.
```

**Comment by mschmidt87** [comment]
Thanks for the feedback, I included all your suggestions. Please take another look. @jakobj

File: examples/README.rst [hunk 23]
```diff
@@ -0,0 +1,5 @@
+========
+Examples
+========
+
+This folder contains a number of example scripts that demonstrate the different use cases of `hal-cgp`.
```
Comment (jakobj) [inline examples/README.rst:5]:
```suggestion
This folder contains a number of example scripts that demonstrate some use cases of `hal-cgp`.
```

File: examples/example_differential_evo_regression.py [hunk 24]
```diff
@@ -57,91 +73,103 @@ def objective(individual, seed):
...
-        "tournament_size": 1,
-        "n_processes": 1,
-        "k_local_search": 2,
-    }
+pop = cgp.Population(**population_params, genome_params=genome_params)
 
-    evolve_params = {"max_generations": 2000, "min_fitness": 0.0}
+# define the function for local search; parameters such as the
+# learning rate and number of gradient steps are fixed via the use
+# of `partial`; the local_search function should only receive a
```
Comment (jakobj) [inline examples/example_differential_evo_regression.py:114]:
```suggestion
# of `partial`; the `local_search` function should only receive a
```

File: examples/example_evo_regression.py [hunk 25]
```diff
@@ -64,6 +81,19 @@ def objective(individual, target_function, seed):
...
 
+# %%
+# Next, we define the main loop of the evolution. To easily execute it
+# for different target functions, we wrap it into a function here. It
+# comprises:
+#
+# - defining the parameters for the population, the genome of individuals,
+#   and the evolutionary algorithm.
+# - creating a Population instance and instantiate the evolutionary algorithm.
+# - we define a recording callback closure for bookkeeping of the progression of the evolution.
```
Comment (jakobj) [inline examples/example_evo_regression.py:92]:
```suggestion
# - creating a Population instance and instantiating the evolutionary algorithm.
# - defining a recording callback closure for bookkeeping of the progression of the evolution.
```

File: examples/example_mountain_car.py [hunk 26]
```diff
@@ -1,3 +1,18 @@
+"""
+Example: Solving an OpenAI Gym environment with CGP.
+====================================================
+
+This examples demonstrates how to solve an OpenAI Gym environment
+(https://gym.openai.com/envs/) with Cartesian genetic programming. We
+choose the MountainCarContinuous environment due to its continuous
```
Comment (jakobj) [inline examples/example_mountain_car.py:7]:
```suggestion
choose the "MountainCarContinuous" environment due to its continuous
```

File: examples/example_mountain_car.py [hunk 27]
```diff
@@ -97,6 +119,16 @@ def objective(ind, seed, n_total_steps):
     return ind
 
 
+# %%
+# We then define the main loop for the evolution, which consists of:
+#
+# - parameters for the population, the genome of individuals, and the evolutionary algorithm.
+# - creating a Population instance and instantiate the evolutionary algorithm.
+# - we define a recording callback closure for bookkeeping of the progression of the evolution.
```
Comment (jakobj) [inline examples/example_mountain_car.py:127]:
```suggestion
# - creating a Population instance and instantiating the evolutionary algorithm.
# - defining a recording callback closure for bookkeeping of the progression of the evolution.
```

File: examples/README.rst (hunk 23 above)
Comment (mschmidt87) [reply examples/README.rst:5]: thanks, I used "several" in the end.

**Comment by mschmidt87** [comment]
Thanks for the feedback, I incorporated your suggestions. Please take another look. @jakobj

**Pull request merged**